pkg/controller/network_policy.go

```diff
@@ -795,71 +811,138 @@ func (cont *AciController) getServiceAugmentBySubnet(subj apicapi.ApicObject,
-					continue
-				}
+					if incomplete {
+						continue
+					}
 
-				proto := portProto(&foundEpPort.Protocol)
-				port := strconv.Itoa(int(svcPort.Port))
-				updateServiceAugmentForService(portAugments,
-					proto, port, service)
+					proto := portProto(&foundEpPort.Protocol)
+					port := strconv.Itoa(int(svcPort.Port))
+					updateServiceAugmentForService(portAugments,
+						proto, port, service)
 
-				logger.WithFields(logrus.Fields{
-					"proto":   proto,
-					"port":    port,
-					"service": servicekey,
-				}).Debug("Allowing egress for service by subnet match")
+					logger.WithFields(logrus.Fields{
+						"proto":   proto,
+						"port":    port,
+						"service": servicekey,
+					}).Debug("Allowing egress for service by subnet match")
+				}
 			}
 		}
+	} else {
+		for servicekey := range matchedServices {
+			serviceobj, _, err := cont.serviceIndexer.GetByKey(servicekey)
+			if err != nil {
+				logger.Error("Could not lookup service for "+
+					servicekey+": ", err.Error())
+				continue
+			}
+			if serviceobj == nil {
+				continue
+			}
+			service := serviceobj.(*v1.Service)
+			label := map[string]string{"kubernetes.io/service-name": service.ObjectMeta.Name}
+			selector := labels.SelectorFromSet(labels.Set(label))
+			cache.ListAllByNamespace(cont.endpointSliceIndexer, service.ObjectMeta.Namespace, selector,
+				func(endpointSliceobj interface{}) {
+					endpointSlices := endpointSliceobj.(*v1beta.EndpointSlice)
+					for _, svcPort := range service.Spec.Ports {
+						if prs.port != nil &&
+							(svcPort.Protocol != *prs.port.Protocol ||
+								svcPort.TargetPort.String() !=
+									prs.port.Port.String()) {
+							// egress rule does not match service target port
+							continue
+						}
+						var foundEpPort *v1beta.EndpointPort
+						for _, endpointPort := range endpointSlices.Ports {
+
+							if *endpointPort.Name == svcPort.Name ||
+								(len(service.Spec.Ports) == 1 &&
+									*endpointPort.Name == "") {
+								foundEpPort = &endpointPort
+								cont.log.Debug("Found EpPort: ", foundEpPort)
+								break
+							}
+							if foundEpPort == nil {
+								continue
+							}
+						}
+						if foundEpPort == nil {
+							return
+						}
+						// @TODO need to revisit this code to check for non ready address
+						incomplete := false
+						for _, endpoint := range endpointSlices.Endpoints {
+							incomplete = incomplete || !checkEndpointslices(subnetIndex, endpoint.Addresses)
+						}
+						if incomplete {
+							continue
+						}
+						proto := portProto(foundEpPort.Protocol)
+						port := strconv.Itoa(int(svcPort.Port))
+						cont.log.Debug("updateServiceAugmentForService: ", service)
+						updateServiceAugmentForService(portAugments,
+							proto, port, service)
+
+						logger.WithFields(logrus.Fields{
+							"proto":   proto,
+							"port":    port,
+							"service": servicekey,
+						}).Debug("Allowing egress for service by subnet match")
+					}
+				})
+		}
```

Comment:
Can we move this code into a function? Possibly collapse both cases (endpoint slice and endpoint) into one routine.

Reply:
@jojimt  Pls take a look fron Cloud perspective (ex: populating Region/Zone information)

Reply:
Sure Kishore will take a look